### PR TITLE
Add other flagship instruction dataset builders

### PR DESCRIPTION
**Pull request by SLR722**
#### Context
Based on the RFC https://github.com/pytorch/torchtune/pull/493 and the alpaca dataset refactor PR https://github.com/pytorch/torchtune/pull/520, added 2 other flagship instruction datasets (grammar, samsum) type we'd like to support in torchtune

#### Changelog
- add grammar dataset builder and related unit test 
- add samsum dataset builder and related unit test 
- modify InstructionDataset class to make it more error prone to partially completed column_map
- add more detailed docstrings for column_map setup

#### Test plan
Added unit test
- pytest tests/torchtune/datasets/test_grammar_dataset.py
- pytest tests/torchtune/datasets/test_samsum_dataset.py

E2E test that kick-off finetune training with new added datasets 
- grammar dataset 
Screenshot 2024-03-20 at 7 09 49 PM
- samsum dataset
Screenshot 2024-03-20 at 7 09 31 PM

#### Discussion item
- The first time using grammar dataset to kick off training need long time (> 1h) to generate training split
Screenshot 2024-03-20 at 7 09 49 PM

### CODE REVIEW

File: torchtune/datasets/_samsum.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from torchtune.data import SummarizeTemplate
+from torchtune.datasets._instruct import InstructDataset
+from torchtune.modules import Tokenizer
+
+
+def samsum_dataset(
+    tokenizer: Tokenizer,
+    train_on_input: bool = True,
```
Comment (RdoubleA) [inline torchtune/datasets/_samsum.py:14]: I think by default we want to leave train_on_input on False, as I've seen most datasets do mask out the prompt when computing loss (except alpaca)

File: torchtune/datasets/_grammar.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from torchtune.data import GrammarErrorCorrectionTemplate
+from torchtune.datasets._instruct import InstructDataset
+from torchtune.modules import Tokenizer
+
+
+def grammar_dataset(
+    tokenizer: Tokenizer,
+    train_on_input: bool = True,
```
Comment (RdoubleA) [inline torchtune/datasets/_grammar.py:14]: same comment below

**Review by RdoubleA — APPROVED** [review]
wow I don't have much to comment except for some minor fixes, this looks super clean. thanks again for adding two additional datasets for two new tasks!
